providers/dns/inwx/inwx.go

```diff
@@ -204,27 +204,34 @@ func (d *DNSProvider) twoFactorAuth(info *goinwx.LoginResponse) error {
 	}
 
 	// INWX forbids re-authentication with a previously used TAN.
-	// To avoid using the same TAN twice, we wait until the next
-	// TOTP period and retry.
-	const s = 30 // seconds per TOTP period
-	now := time.Now()
-	tanCounter := now.Unix() / s
-	if tanCounter == d.prevTANCounter {
-		nextPeriod := time.Unix(s*(now.Unix()/s)+s, 0)
-		d := nextPeriod.Sub(now)
-
-		log.Infof("inwx: waiting %v for next TOTP token", d)
-		time.Sleep(d)
-
-		now = time.Now()
-		tanCounter = now.Unix() / s
+	// To avoid using the same TAN twice, we wait until the next TOTP period and retry.
+	sleep := d.computeSleep(time.Now())
+	if sleep != 0 {
+		log.Infof("inwx: waiting %s for next TOTP token", sleep)
+		time.Sleep(sleep)
 	}
 
-	tan, err := totp.GenerateCode(d.config.SharedSecret, time.Now())
+	now := time.Now()
+
+	tan, err := totp.GenerateCode(d.config.SharedSecret, now)
 	if err != nil {
 		return err
 	}
-	d.prevTANCounter = tanCounter
+
+	d.previousCall = now
 
 	return d.client.Account.Unlock(tan)
 }
+
+func (d *DNSProvider) computeSleep(now time.Time) time.Duration {
+	if d.previousCall.IsZero() {
+		return 0 * time.Second
+	}
+
+	endPeriod := d.previousCall.Add(30 * time.Second)
```

Comment:
While this works, I suspect we are wasting a few seconds here on average, because d.previousCall is never rooted to be a multiple of 30 seconds after the epoch (Unix time 0)?

It works though and how relevant this is depends a bit on the distance between ajacent TOTP attempts.

Reply:
You are right, sorry I completely missed that.
Sorry, I will try to fix that.
I will truncate the previous time by 30 seconds.